elyra/pipeline/component.py

```diff
@@ -194,41 +176,83 @@ def __init__(self, id: str, name: str, description: Optional[str], runtime: Opti
         self._id = id
         self._name = name
         self._description = description
+        self._source_type = source_type
+        self._source = source
+
         self._runtime = runtime
-        self._properties = properties
         self._op = op
-        self._extension = extension
+        self._category_id = category_id
+        self._properties = properties
+
+        if not parameter_refs:
+            if self._source_type == "elyra":
+                parameter_refs = {
+                    "filehandler": "filename"
+                }
+            else:
+                parameter_refs = {}
+
+        if extensions and not parameter_refs.get('filehandler'):
+            Component._log_warning(f"Component '{self._id}' specifies extensions '{extensions}' but \
+                                   no entry in the 'parameter_ref' dictionary for 'filehandler' and \
+                                   cannot participate in drag and drop functionality as a result.")
+
+        self._extensions = extensions
+        self._parameter_refs = parameter_refs
 
     @property
-    def id(self):
+    def id(self) -> str:
         return self._id
 
     @property
-    def name(self):
+    def name(self) -> str:
         return self._name
 
     @property
-    def description(self):
+    def description(self) -> Optional[str]:
         return self._description
 
     @property
-    def runtime(self):
-        return self._runtime
+    def source_type(self) -> str:
+        return self._source_type
 
     @property
-    def properties(self):
-        return self._properties
+    def source(self) -> str:
+        return self._source
 
     @property
-    def op(self):
+    def runtime(self) -> Optional[str]:
+        return self._runtime
+
+    @property
+    def op(self) -> Optional[str]:
         if self._op:
             return self._op
         else:
             return self._id
 
     @property
-    def extension(self):
-        return self._extension
+    def category_id(self) -> Optional[str]:
```

Comment:
Is category really optional? Can we have a component without a category? Would the UI break in this case?

Reply:
No idea.  This was listed as `Optional` because the constructor uses `None` as a default value.  The effect on the UI was probably not attempted.